example/scripts/lock-stack.js

```diff
@@ -3,48 +3,160 @@
 'use strict';
 
 /**
- * Functionality for locking stacks with expiration using the Mutex provided by
- * @cumulus/common. Stacks are locked based on the deployment name.
+ * Functionality for locking stacks with expiration and SHA tracking
  *
  * To use this, a DynamoDB table is required with the primary partition key being
  * a string - 'key' and no sort key. The table name should be set in LOCK_TABLE_NAME
  */
 
-const {
-  concurrency
-} = require('@cumulus/common');
-const aws = require('@cumulus/common/aws');
+const { aws } = require('@cumulus/common');
 
-const LOCK_TABLE_NAME = 'cumulus-int-test-lock';
-const STACK_EXPIRATION_MS = 120 * 60 * 1000; // 2 hours
+class CumulusLockError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+  }
+}
 
-function performLock(mutex, deployment) {
-  return mutex.writeLock(deployment, STACK_EXPIRATION_MS);
+
+class CumulusNoLockError extends Error {
+  constructor(message) {
+    super(message);
+    this.name = this.constructor.name;
+  }
 }
 
-function removeLock(mutex, deployment) {
-  return mutex.unlock(deployment);
+class Mutex {
+  constructor(docClient, tableName) {
+    this.docClient = docClient;
+    this.tableName = tableName;
+  }
+
+  async checkMatchingSha(key, gitSHA) {
+    const params = {
+      TableName: this.tableName,
+      Key: {
+        key: key
+      }
+    };
+    const record = await this.docClient.get(params).promise();
+    if (record.Item) {
+      return (gitSHA === record.Item.sha ? 'match' : record.Item.sha);
+    }
+    return 'noLock';
+  }
+
+  async writeLock(key, timeoutMs, gitSHA) {
+    const now = Date.now();
+    const params = {
+      TableName: this.tableName,
+      Item: {
+        key: key,
```

Comment:
So just to make sure I understand, the `key` is the deployment name and you can have different locks for different SHAs? Are `key` and `sha` both part of the primary index (or key or whatever it's called in Dynamo)? Otherwise wouldn't writing a newer value for the same `key` overwrite a previous one?

Reply:
No - each deployment can have *one* lock, the `sha` attribute is set at lock time, and that attribute must be provided to unlock the stack.      

You *can* have the same key overwrite a previous one, regardless of the sha value *if* the lock timer has expired, that's desired behavior (and currently as implemented) , as it prevents CI failures  from forever locking a stack without direct user cleanup.